### ISSUE TITLE
Issue verifying signature Subscription

### ISSUE THREAD

**Issue by ggrandes** [aws/aws-sdk-java#635]
Acording to AWS-SNS [Verifying Message Signatures Doc](http://docs.aws.amazon.com/sns/latest/dg/SendMessageToHttp.verify.signature.html):

SubscriptionConfirmation and UnsubscribeConfirmation tokens are:

    Message
    MessageId
    SubscribeURL
    Timestamp
    Token
    TopicArn
    Type

[com.amazonaws.services.sns.util.SignatureChecker](https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-sns/src/main/java/com/amazonaws/services/sns/util/SignatureChecker.java):

```java
private TreeMap<String, String> subscribeMessageValues(Map<String, String> parsedMessage){
    TreeMap<String, String> signables = new TreeMap<String, String>();
    String[] keys = { SUBSCRIBE_URL, MESSAGE, MESSAGE_ID, TYPE, TIMESTAMP, TOKEN, TOPIC };
    for(String key: keys){
        if(parsedMessage.containsKey(key)){
            signables.put(key, parsedMessage.get(key));
        }
    }
    return signables;
}
```

Correct order must be:

```java
    String[] keys = { MESSAGE, MESSAGE_ID, SUBSCRIBE_URL, TIMESTAMP, TOKEN, TOPIC, TYPE };
```

**Comment by manikandanrs** [comment]
Are you seeing a signature issue without this ? TreeMap takes care of ordering based on the keys.

**Comment by ggrandes** [comment]
Well, you are right, order of keys are irrelevant (The change I propose don't have any effect). 
I have an issue verifying Subscription signature (but verifying Notification works good). I'm using SNS HTTP endpoint (a simple Servlet), in both cases I use:

```java
final SignatureChecker signatureChecker = new SignatureChecker();
return signatureChecker.verifyMessageSignature(source, pubKey);
```

**Comment by ggrandes** [comment]
I have found the problem... missing SUBSCRIBE_URL in INTERESTING_FIELDS:

```java
--- a/aws-java-sdk-sns/src/main/java/com/amazonaws/services/sns/util/SignatureChecker.java
+++ b/aws-java-sdk-sns/src/main/java/com/amazonaws/services/sns/util/SignatureChecker.java
@@ -56,7 +56,7 @@ public class SignatureChecker {
     private final String TOPIC = "TopicArn";
     private final String TOKEN = "Token";
     private final Set<String> INTERESTING_FIELDS = new HashSet<String>(Arrays.asList(TYPE, SUBSCRIBE_TYPE, MESSAGE,
-            TIMESTAMP, SIGNATURE, SIGNATURE_VERSION, MESSAGE_ID, SUBJECT, TOPIC, TOKEN));
+            TIMESTAMP, SIGNATURE, SIGNATURE_VERSION, MESSAGE_ID, SUBJECT, TOPIC, TOKEN, SUBSCRIBE_URL));
 
     /**
      * Validates the signature on a Simple Notification Service message. No
```

**Comment by ggrandes** [comment]
This bug was introduced in version 1.10.28 [commit 8482a46bd635cd42a10ffd4d44e0fe4937fa9db9] and was fixed in [Pull Request #594], but seems a regression...

**Comment by manikandanrs** [comment]
Thanks. I will get this fixed.

**Comment by manikandanrs** [comment]
The fix was released as part of yesterday's release 1.10.53.

**Comment by ggrandes** [comment]
Thanks. Is fixed! :+1:
